payjoin-cli/src/app/v2/mod.rs

```diff
@@ -248,64 +237,183 @@ impl App {
+        proposal: Receiver<OutputsUnknown>,
+        persister: &ReceiverPersister,
+    ) -> Result<()> {
+        let wallet = self.wallet();
+        let proposal = proposal
+            .identify_receiver_outputs(|output_script| Ok(wallet.is_mine(output_script)?))
+            .save(persister)?;
+        self.commit_outputs(proposal, persister).await
+    }
 
-        // Receive Check 2: receiver can't sign for proposal inputs
-        let proposal = proposal.check_inputs_not_owned(|input| Ok(wallet.is_mine(input)?))?;
-        log::trace!("check2");
+    async fn commit_outputs(
+        &self,
+        proposal: Receiver<WantsOutputs>,
+        persister: &ReceiverPersister,
+    ) -> Result<()> {
+        let proposal = proposal.commit_outputs().save(persister)?;
+        self.contribute_inputs(proposal, persister).await
+    }
 
-        // Receive Check 3: have we seen this input before? More of a check for non-interactive i.e. payment processor receivers.
-        let payjoin = proposal
-            .check_no_inputs_seen_before(|input| Ok(self.db.insert_input_seen_before(*input)?))?;
-        log::trace!("check3");
+    async fn contribute_inputs(
+        &self,
+        proposal: Receiver<WantsInputs>,
+        persister: &ReceiverPersister,
+    ) -> Result<()> {
+        let wallet = self.wallet();
+        let candidate_inputs = wallet.list_unspent()?;
 
-        let payjoin = payjoin
-            .identify_receiver_outputs(|output_script| Ok(wallet.is_mine(output_script)?))?
-            .commit_outputs();
+        let selected_input = proposal.try_preserving_privacy(candidate_inputs)?;
+        let proposal =
+            proposal.contribute_inputs(vec![selected_input])?.commit_inputs().save(persister)?;
+        self.finalize_proposal(proposal, persister).await
+    }
 
-        let provisional_payjoin = try_contributing_inputs(payjoin.clone(), &wallet)
-            .map_err(ReplyableError::Implementation)?;
+    async fn finalize_proposal(
+        &self,
+        proposal: Receiver<ProvisionalProposal>,
+        persister: &ReceiverPersister,
+    ) -> Result<()> {
+        let wallet = self.wallet();
+        let proposal = proposal
+            .finalize_proposal(
+                |psbt| Ok(wallet.process_psbt(psbt)?),
+                None,
+                self.config.max_fee_rate,
+            )
+            .save(persister)?;
+        self.send_payjoin_proposal(proposal, persister).await
+    }
 
-        let payjoin_proposal = provisional_payjoin.finalize_proposal(
-            |psbt| Ok(wallet.process_psbt(psbt)?),
-            None,
-            self.config.max_fee_rate,
-        )?;
-        let payjoin_proposal_psbt = payjoin_proposal.psbt();
-        log::debug!("Receiver's Payjoin proposal PSBT Rsponse: {payjoin_proposal_psbt:#?}");
-        Ok(payjoin_proposal)
+    async fn send_payjoin_proposal(
+        &self,
+        mut proposal: Receiver<PayjoinProposal>,
+        persister: &ReceiverPersister,
```

Comment:
It sure seems like a lot of these functions require &ReceiverPersister. Should it not be part of `self`, either as an application field or a sub Receiver field?

Reply:
Since the session persister is bound to a specific session would this work as a application param? IIRC resume spawns tokio "threads" for each available session

Reply:
gotcha, resolved then.